fem/libceed/ceed.cpp

```diff
@@ -382,7 +578,7 @@ void InitCeedBasisAndRestriction(const FiniteElementSpace &fes,
       {
          InitCeedNonTensorRestriction(fes, irm, ceed, restr);
       }
-      internal::ceed_restr_map[restr_key] = *restr;
+      // internal::ceed_restr_map[restr_key] = *restr;
```

Comment:
Why is this disabled?

Reply:
Using the address of the pointer as a hash key creates bugs, @jeremylt is looking into a more robust key.

Reply:
Or we need to find a way to remove the entries in the map when we destroy a `FiniteElementSpace`.

Reply:
Okay, I pushed a fix, tell me if you like it.

Reply:
I added a `RemoveCeedBasisAndRestriction` function that is called in the destructor of `FiniteElementSpace`.

Reply:
This will be a bigger change (maybe for another PR) but I think we can get rid of the global containers and instead associate the ceed restrictions with the `FiniteElementSpace` objects and the ceed bases with the `FiniteElement` objects. This is what we do for the mfem versions of the restriction and "basis" (aka `DofToQuad`) objects.

Reply:
I like the idea, I think @jeremylt had something like this in mind. I can do it in an other PR.

Reply:
Does a FE object hold both a FES and a IR?

This is the sort of information I was talking about on slack that would be nice to have in some developer documentation.

Reply:
@jeremylt, here are some notes on that.

The `FiniteElement` objects do not know about `FiniteElementSpace` and they does not own any `IntegrationRule` objects.

`FiniteElementSpace` objects have a pointer to a `FiniteElementColloection`, which they do not own, and which can be queried for `FiniteElement` objects, e.g. based on the element types present in the mesh. Also, `FiniteElementSpace` has a pointer to a `Mesh` which it does not own.

`FiniteElement` objects can be asked for a "basis" description (aka `DofToQuad`) for a given quadrature rule using: https://github.com/mfem/mfem/blob/a40523a7cfe6bb0037311cd46200e2a5eddfc180/fem/fe.hpp#L561-L565

The returned `DofToQuad` object is lazy-constructed, cached, and owned by the `FiniteElement`.

Restriction operators are similarly handled by `FiniteElementSpace` objects using the method: https://github.com/mfem/mfem/blob/a40523a7cfe6bb0037311cd46200e2a5eddfc180/fem/fespace.hpp#L337-L354